internal/server/data/data.go

```diff
@@ -152,28 +139,62 @@ func add[T models.Modelable](db *gorm.DB, model *T) error {
 		return err
 	}
 
-	if err := db.Create(model).Error; err != nil {
-		if isUniqueConstraintViolation(err) {
-			return fmt.Errorf("%w: %s", internal.ErrDuplicate, err)
-		}
+	err := db.Create(model).Error
+	return handleError(err)
+}
 
-		return err
-	}
+type UniqueConstraintError struct {
+	Table  string
+	Column string
+}
 
-	return nil
+func (e UniqueConstraintError) Error() string {
+	return fmt.Sprintf("value for %v already exists in %v", e.Column, e.Table)
 }
 
-func isUniqueConstraintViolation(err error) bool {
-	var pgerr *pgconn.PgError
-	if errors.As(err, &pgerr) {
-		return pgerr.Code == pgerrcode.UniqueViolation
+// handleError looks for well known DB errors. If the error is recognized it
+// is translated into a UniqueConstraintError so that calling code can
+// inspect the error.
+func handleError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) {
+		switch pgErr.Code {
+		case pgerrcode.UniqueViolation:
+			// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/access/nbtree/nbtinsert.c;h=f6f4af8bfe3f0876e860944e2233328f8cc3303a;hb=HEAD#l668
+			// ereport(ERROR,
+			//         (errcode(ERRCODE_UNIQUE_VIOLATION),
+			//          errmsg("duplicate key value violates unique constraint \"%s\"",
+			//                 RelationGetRelationName(rel)),
+			//                 key_desc ? errdetail("Key %s already exists.",
+			//                            key_desc) : 0,
+			//                 errtableconstraint(heapRel,
+			//                                    RelationGetRelationName(rel))));
+			re := regexp.MustCompile(`Key \((?P<columnName>\w+)\)=\(\w+\) already exists.`)
+			matches := re.FindStringSubmatch(pgErr.Detail)
+			if matches != nil {
+				columnNameIndex := re.SubexpIndex("columnName")
+				return UniqueConstraintError{Table: pgErr.TableName, Column: matches[columnNameIndex]}
```

Comment:
While there is a `ColumnName` field in the error struct, it's no populated for this particular error

Reply:
Oh wow, that's surprising. Is it the `ConstraintName` that is set? Which maybe is not that helpful for us.

Reply:
I was surprised as well. This is the full error from psql
```
postgres      | 2022-05-30 18:15:11.405 UTC [35] ERROR:  duplicate key value violates unique constraint "idx_credentials_identity_id"
postgres      | 2022-05-30 18:15:11.405 UTC [35] DETAIL:  Key (identity_id)=(54270808554807296) already exists.
postgres      | 2022-05-30 18:15:11.405 UTC [35] STATEMENT:  INSERT INTO "credentials" ("created_at","updated_at","deleted_at","identity_id","password_hash","one_time_password","one_time_password_used","id") VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING "id"
```
and seen from infra
```
infra-server  | {"level":"info","ts":1653934511.4067185,"caller":"data/data.go:165","msg":"&pgconn.PgError{Severity:\"ERROR\", Code:\"23505\", Message:\"duplicate key value violates unique constraint \\\"idx_credentials_identity_id\\\"\", Detail:\"Key (identity_id)=(54270808554807296) already exists.\", Hint:\"\", Position:0, InternalPosition:0, InternalQuery:\"\", Where:\"\", SchemaName:\"public\", TableName:\"credentials\", ColumnName:\"\", DataTypeName:\"\", ConstraintName:\"idx_credentials_identity_id\", File:\"nbtinsert.c\", Line:663, Routine:\"_bt_check_unique\"}"}
```

`ConstraintName` is populated but we'd still need to parse the constraint string

Reply:
That's a good idea. I'll update it to use the constraint name instead. This would only be applicable to postgres. AFAIK, sqlite still requires parsing